### DISCUSSION TITLE
PROBLEM ON SAVING A PARTICULAR DATA TYPE

### DISCUSSION THREAD

**Discussion by cugwu** [Project-MONAI/tutorials#1234] [Q&A] [answered]
Hi to everyone, I'm trying to save my predictions as .nii.gz of data type uint8. I'm using:
`            transforms.SaveImaged(keys="pred", meta_keys="pred_meta_dict", output_dir=args.out_dir, output_ext=".nii.gz",
                                  output_dtype=np.uint8, output_postfix="", separate_folder=False, resample=False),`

but even if I put **output_dtype=np.uint8** my predictions are still saved as np.float32. As a double check, I try to do the opposite and see if I load an image using 
`LoadImaged(keys=["image", "label"], dtype=np.uint8),`
and forcing the data type to be uint8 will work but it doesn't work. No matter the datatype I put the default one, which is np.float32, will be used. 

I've also tried writing torch.uint8, and "uint8". But it is not working. 
Can someone please help me with this problem?

Thanks in advance for the answer.

**Comment by ericspod** [comment, marked as answer]
What version of MONAI are you using? Please run `monai.config.print_config()` to print out the config information. I'm using the current dev version of MONAI which is based off 1.1 and am getting these results:

```python
import torch
import numpy as np
import nibabel as nib
from monai.utils import ImageMetaKey
from monai.data import MetaTensor
from monai.transforms import SaveImage

m=MetaTensor(torch.zeros(16,16,dtype=torch.uint8))
m.meta[ImageMetaKey.FILENAME_OR_OBJ]="test"

s=SaveImage(output_dtype=np.uint8, output_dir="/tmp", output_ext=".nii.gz", output_postfix="", separate_folder=False, resample=False)
s(m)

d=nib.load("/tmp/test.nii.gz")
print(d.dataobj.dtype)  # prints uint8
```

The nifti file here appears to be saving uint8 data correctly, the transform isn't dictionary based like yours but it should be equivalent code.

**Reply by cugwu** [reply]
Hi my monai configuration is:
MONAI version: 1.1.0
Numpy version: 1.24.1

If I'm using monai.LoadImaged(keys=["image", "label"], dtype=None) that should load the data with the original data type I keep getting float32

If a load the mage with nibabel I get uint8. 

I'm still not understanding what is happening but thanks for you help

**Reply by ericspod** [reply]
```python
from monai.transforms import  LoadImage

l=LoadImage(image_only=True, dtype=None)
d=l("/tmp/test.nii.gz")
print(d.dtype)  # prints uint8
```

This works in that the loaded dtype is uint8. I don't think again using dictionary transforms makes a difference looking at the code for `LoadImaged`. Does this work for you?

**Reply by cugwu** [reply]
yes`LoadImage` works (output: torch.uint8). In the dictionary-based approach, I removed the other post-processing transformation and now with only  `LoadImaged` the data type is uint8. probably one of the other post-transformations is modifying my data type. Thanks for the help!

**Reply by ericspod** [reply]
No problem, I know that Pytorch has issues with uint8 so in places there is likely code doing an implicit conversion because it's necessary.
